metadata/src/main/java/com/google/cloud/teleport/metadata/auto/AutoTemplate.java

```diff
@@ -0,0 +1,318 @@
+              options.as(sourceBlockInstance.getOptionsClass()));
+
+      for (ExecutionBlock executionBlock : orderedBlocks) {
+        TemplateBlock blockInstance = executionBlock.blockInstance;
+        PipelineOptions optionsClass = options.as(blockInstance.getOptionsClass());
+
+        input =
+            executionBlock.blockMethod.invoke(executionBlock.blockInstance, input, optionsClass);
+
+        if (dlqInstance != null) {
+          ExecutionBlock dlqBlock;
+          if (dlqOutputs(executionBlock.getBlockMethod()) != null) {
+            dlqBlock =
+                buildDlqExecutionBlock(templateClass, dlqOutputs(executionBlock.getBlockMethod()));
+            dlqBlock.blockMethod.invoke(
+                dlqBlock.blockInstance, input, options.as(dlqInstance.getOptionsClass()));
+          }
+        }
+      }
+
+      pipeline.run();
+
+    } catch (Exception e) {
+      throw new RuntimeException("Error instantiating source", e);
+    }
+  }
+
+  public static List<ExecutionBlock> buildExecutionBlocks(Class<?> templateClass)
+      throws InstantiationException,
+          IllegalAccessException,
+          InvocationTargetException,
+          NoSuchMethodException {
+
+    Template annotations = templateClass.getAnnotation(Template.class);
+    if (annotations == null) {
+      throw new IllegalStateException(
+          "Class "
+              + templateClass
+              + " does not have a @Template annotation, can not use auto template features.");
+    }
+    Class<?>[] blocks = annotations.blocks();
+    if (blocks.length == 0) {
+      throw new IllegalStateException(
+          "Class "
+              + templateClass
+              + " does not have a @Template annotation with valid blocks, can not use auto template features.");
+    }
+
+    List<Class<?>> chainedClasses = Arrays.asList(blocks);
+    LOG.info("Chaining blocks {}", chainedClasses);
+
+    Class<? extends TemplateTransform> sourceBlockClass =
+        chainedClasses.get(0).asSubclass(TemplateTransform.class);
+    TemplateTransform<?> templateSourceInstance =
+        sourceBlockClass.getDeclaredConstructor().newInstance();
+
+    Method sourceMethod = getReadMethod(sourceBlockClass);
+    LOG.info("Going to read source method {}", sourceMethod);
+
+    List<Class<?>> transformations = chainedClasses.subList(1, chainedClasses.size() - 1);
+    LOG.info("Transformations are {}", transformations);
+
+    Outputs previousType = outputs(sourceMethod);
+
+    List<ExecutionBlock> orderedBlocks = new ArrayList<>();
+    orderedBlocks.add(new ExecutionBlock(sourceBlockClass, templateSourceInstance, sourceMethod));
+
+    for (Class<?> transformationClass : transformations) {
+
+      LOG.info("Returned type from the previous block: {}", previousType);
+
+      Method transformMethod =
+          getTransformMethod(transformationClass, previousType.value(), previousType.types());
+
+      LOG.info("Next method will be {}", transformMethod);
```

Comment:
nit Those should likely become .debug when we submit

Reply:
Done

Reply:
Done